### PR TITLE
Migration to scaffold framework for code-generation

### LINKED ISSUE #58: Can't run SQL against MS Access (unable to obtain a DbFactory)
I posted this on the LinqPad forum as well, because I'm not sure if it's a limitation of LinqPad or an issue with this driver.

When I make a connection to an MS Access db (either OleDb or ODBC), I'm able to test the connection successfully, I'm able to expand the connection node and see all the tables, and I'm able to run C# code using the connection. But if I choose SQL as the language and try to execute a straight query, I get an error, "Unable to obtain a DbFactory for the connection."

### LINKED ISSUE #78: Migrate to new scaffold framework for model generation from old T4-based code

### LINKED ISSUE #79: Add ClickHouse support

### PR DESCRIPTION
**Pull request by MaceWindu**
Fix #58 
Fix #78
Fix #79

Major driver rework. Check release-notes.md file for all changes

Main changes:
- migration to new scaffold framework
- rework of settings dialog
- bump .net framework requirements for LINQPad 5 driver to .NET Framework 4.8.0 to get rid of old dependencies and improve compatibility with providers without old netfx versions support
- ~re-introduce DB2 iSeries support~ not today, postponed
- add ClickHouse support

Note that there are several issues on linq2db side regarding schema and scaffold functionality for some providers, which will be fixed later

### CODE REVIEW

File: linq2db.LINQPad.sln [hunk 1]
```diff
@@ -27,13 +20,23 @@ Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "Build", "Build", "{A26F7937
...
 	EndProjectSection
 EndProject
+Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "Templates", "Templates", "{B3100892-7A58-4B4E-A926-0D5FE4309C4A}"
+	ProjectSection(SolutionItems) = preProject
+		LinqToDB.Templates\NotifyPropertyChanged.ttinclude = LinqToDB.Templates\NotifyPropertyChanged.ttinclude
+		LinqToDB.Templates\README.md = LinqToDB.Templates\README.md
+		LinqToDB.Templates\T4Model.ttinclude = LinqToDB.Templates\T4Model.ttinclude
+	EndProjectSection
+EndProject
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "LinqToDB.Tools", "..\linq2db\Source\LinqToDB.Tools\LinqToDB.Tools.csproj", "{91DE5A8B-4DF8-4E27-A33B-886563CF7612}"
```
Comment (MaceWindu) [inline linq2db.LINQPad.sln:38]: temporary refenrece to code from https://github.com/linq2db/linq2db/pull/3803

File: release-notes.md [hunk 2]
```diff
@@ -1,3 +1,35 @@
...
+- [ALL] dependency update: Microsoft.CodeAnalysis.CSharp updated to 4.3.1
+- [ALL] dependency update: Npgsql updated to 6.0.7
+- [LINQPAD7] add new dependency: Octonica.ClickHouseClient 2.2.9
+- [LINQPAD5] dependency update: Oracle.ManagedDataAccess 19.14.0 -> 21.7.0
+- [LINQPAD5] dependency update: Microsoft.SqlServer.Types 14.0.1016.290 -> 160.900.6-rc0
+- [LINQPAD7] dependency update: FirebirdSql.Data.FirebirdClient 9.0.1 -> 9.0.2
+- [LINQPAD7] dependency update: dotMorten.Microsoft.SqlServer.Types 1.5.0 -> 2.5.0
+- [LINQPAD7] dependency update: FirebirdSql.Data.FirebirdClient 9.0.1 -> 9.0.2
+
+# Release 4.2.0
```
Comment (MaceWindu) [inline release-notes.md:25]: missing 4.2.0 release notes

File: Source/Compat/IsExternalInit.cs [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+using System.ComponentModel;
+
+namespace System.Runtime.CompilerServices;
+
+/// <summary>
+/// Reserved to be used by the compiler for tracking metadata.
+/// This class should not be used by developers in source code.
+/// </summary>
+[EditorBrowsable(EditorBrowsableState.Never)]
+internal static class IsExternalInit
```
Comment (viceroypenguin) [inline Source/Compat/IsExternalInit.cs:16]: Should use [PolySharp](https://www.nuget.org/packages/PolySharp) for this. It's done via code-generation, so it does not add a runtime dependency.

File: Source/Compat/NullableAttributes.cs [hunk 4]
```diff
@@ -0,0 +1,151 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+// Rationale: required for nullable analysis with older frameworks without those attributes
+// Source: https://github.com/dotnet/runtime/blob/master/src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/NullableAttributes.cs
+namespace System.Diagnostics.CodeAnalysis;
```
Comment (viceroypenguin) [inline Source/Compat/NullableAttributes.cs:7]: Also PolySharp

File: Source/Compat/SkipLocalsInitAttribute.cs [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+/// </remarks>
+[AttributeUsage(AttributeTargets.Module
+	| AttributeTargets.Class
+	| AttributeTargets.Struct
+	| AttributeTargets.Interface
+	| AttributeTargets.Constructor
+	| AttributeTargets.Method
+	| AttributeTargets.Property
+	| AttributeTargets.Event, Inherited = false)]
+internal sealed class SkipLocalsInitAttribute : Attribute
```
Comment (viceroypenguin) [inline Source/Compat/SkipLocalsInitAttribute.cs:28]: Also PolySharp

File: release-notes.md [hunk 6]
```diff
@@ -1,3 +1,51 @@
...
+- implement automatic schema refresh for Access (OLE DB provider only), ClickHouse, DB2 LUW, DB2 z/OS, MySql, MariaDB, SAP HANA, Oracle and SAP/Sybase ASE (previous versions implemented only SQL Server support)
+- added `LinqToDB`, `LinqToDB.Data` and `LinqToDB.Mapping` usings to queries also for static context driver
+- stored procedures and functions schema load options now allow you to select which routine types to load (e.g. only stored procedures)
+- add additional connection string support for Access to mitigate schema issues in ODBC and OLE DB providers
+- show packages in schema tree for supporting databases
+- add nullability information for columns and parameters for both value and reference types
+- add SQL dialog selection for databases for which Linq To DB supports multiple dialects
+- [LINQPAD5] rise required framework version from .NET Framework 4.6.1 to 4.8
+- [LINQPAD7] rise required framework version from `netcoreapp3.1` to `net6.0-windows`
```
Comment (viceroypenguin) [inline release-notes.md:17]:
```suggestion
- [LINQPAD7] raise required framework version from `netcoreapp3.1` to `net6.0-windows`
```

**Review by viceroypenguin — COMMENTED** [review]
Preliminary infra review - I haven't read any of the code yet.

**Review by igor-tkachev — COMMENTED** [review]
Provider names for ClickHouse are confusing. Add Octonica/ClickHouse.Client suffix.

**Review by igor-tkachev — COMMENTED** [review]
ClickHouse Octonica provider is not working for plain SQL.

Switch Language to SQL and execute

```sql
select version()
```

_Specified method is not supported._

File: README.md [hunk 7]
```diff
@@ -1,37 +1,37 @@
-## LINQ to DB LINQPad Driver
+# LINQ to DB LINQPad Driver
 
 [![NuGet Version and Downloads count](https://buildstats.info/nuget/linq2db.LINQPad?includePreReleases=true)](https://www.nuget.org/packages/linq2db.LINQPad) [![License](https://img.shields.io/github/license/linq2db/linq2db.LINQPad)](MIT-LICENSE.txt)
 
 [![Master branch build](https://img.shields.io/azure-devops/build/linq2db/linq2db/8/master?label=build%20(master))](https://dev.azure.com/linq2db/linq2db/_build?definitionId=8&_a=summary) [![Latest build](https://img.shields.io/azure-devops/build/linq2db/linq2db/8?label=build%20(latest))](https://dev.azure.com/linq2db/linq2db/_build?definitionId=8&_a=summary)
 
-linq2db.LINQPad is a driver for [LINQPad 5 (.NET Framework)](http://www.linqpad.net) and [LINQPad 6-7 (.NET Core)](http://www.linqpad.net).
+linq2db.LINQPad is a driver for [LINQPad 5 (.NET Framework 4.8)](http://www.linqpad.net) and [LINQPad 7 (.NET 6+)](http://www.linqpad.net).
```
[the remark this answers is not in the document]
Comment (MaceWindu) [reply README.md:7]: LINQPad 5 is for .NET Framework
LINQPad 6+ are for .NET [Core]

**Comment by MaceWindu** [comment]
> ClickHouse FinedString(20) SQL type is converted to byte[]. Needs to be string.

`ClickHouse` actually doesn't have string data type. `FixedString` (and `String`) are binary types not representable as strings in general. But we can add option to use string for them

> Provider names for ClickHouse are confusing. Add Octonica/ClickHouse.Client suffix.

done

> ClickHouse Octonica provider is not working for plain SQL.

Provider bug. LINQPad tries to save data to DataSet which calls schema API, not implemented by Octonica provider: https://github.com/Octonica/ClickHouseClient/issues/55

There is also another bug with Octonica and raw SQL for queries that doesn't return data (DML/DDL): https://github.com/Octonica/ClickHouseClient/issues/75

BTW, I would recommend to use HTTPS protocol provider as it has much less issues than other ClickHouse providers. And it could be used in .NET Framework too

**Comment by MaceWindu** [comment]
> we can add option to use string for them

option added

File: Source/Compat/IsExternalInit.cs (hunk 3 above)
Comment (MaceWindu) [reply Source/Compat/IsExternalInit.cs]: done

**Review by viceroypenguin — APPROVED** [review]

**Pull request merged**